src/libp2p/connection/established/multi_stream.rs

```diff
@@ -0,0 +1,757 @@
+    /// [`MultiStream::pull_event`] to empty the queue of events between calls to this method.
+    ///
+    /// # Panic
+    ///
+    /// Panics if there is no substream with that identifier.
+    ///
+    pub fn substream_read_write(
+        &mut self,
+        substream_id: &TSubId,
+        read_write: &'_ mut ReadWrite<'_, TNow>,
+    ) -> bool {
+        loop {
+            // Don't process any more data before events are pulled.
+            if self.pending_events.len() >= MAX_PENDING_EVENTS {
+                return false;
+            }
+
+            // TODO: not great to remove then insert back the substream
+            let (substream_id, (mut substream, out_substream_id)) =
+                self.in_substreams.remove_entry(substream_id).unwrap();
+
+            // Reading/writing the ping substream is used to queue new outgoing pings.
+            if Some(&substream_id) == self.ping_substream.as_ref() {
+                if read_write.now >= self.next_ping {
+                    let payload = self
+                        .ping_payload_randomness
+                        .sample(rand::distributions::Standard);
+                    substream.queue_ping(&payload, read_write.now.clone() + self.ping_timeout);
+                    self.next_ping = read_write.now.clone() + self.ping_interval;
+                }
+
+                read_write.wake_up_after(&self.next_ping);
+            }
+
+            let written_bytes_before = read_write.written_bytes;
+
+            let (mut substream_update, event) = substream.read_write(read_write);
+
+            if event.is_none() && read_write.written_bytes == written_bytes_before {
+                self.ready_substreams.remove(&out_substream_id);
+            }
+
+            match (event, substream_update.as_mut()) {
+                (None, _) => {}
+                (Some(substream::Event::InboundNegotiated(protocol)), Some(substream)) => {
+                    if protocol == self.ping_protocol {
+                        substream.set_inbound_ty(substream::InboundTy::Ping);
+                    } else if let Some(protocol_index) = self
+                        .request_protocols
+                        .iter()
+                        .position(|p| p.name == protocol)
+                    {
+                        substream.set_inbound_ty(substream::InboundTy::Request {
+                            protocol_index,
+                            request_max_size: if let ConfigRequestResponseIn::Payload { max_size } =
+                                self.request_protocols[protocol_index].inbound_config
+                            {
+                                Some(max_size)
+                            } else {
+                                None
+                            },
+                        });
+                    } else if let Some(protocol_index) = self
+                        .notifications_protocols
```

Comment:
what's a notification protocol?

Reply:
It's one of the two types of protocols, alongside with request-responses. You open the substream, there's a handshake, then you send notifications. Notification substreams are unidirectional, so if the remote wants to send back notifications it needs to open its own substream.